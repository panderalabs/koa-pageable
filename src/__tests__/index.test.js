// flow
/* eslint-disable no-restricted-syntax */
import {
  Direction,
  ArrayPage,
  IndexablePage,
  IndexedPage,
  Order,
  Pageable,
  Sort,
  NumberFormatError,
  InvalidSortError,
  paginate,
} from '../index';


describe('Tests', () => {
  const orders = [
    new Order('propertyA', Direction.asc),
    new Order('propertyB', Direction.desc),
    new Order('propertyC', Direction.asc),
    new Order('propertyD', Direction.desc),
    new Order('propertyE', Direction.desc),
  ];

  const pageOrders = [
    new Order('firstName', 'asc'),
    new Order('lastName', 'asc'),
  ];

  const content = [
    { id: 1, firstName: 'Bob', lastName: 'Stevens' },
    { id: 2, firstName: 'Steve', lastName: 'Bobbins' },
    { id: 3, firstName: 'Robert', lastName: 'Stevenson' },
    { id: 4, firstName: 'Stevarino', lastName: 'Robertson' },
  ];

  describe('Order class', () => {
    it('returns "asc" for the static _DEFAULT_DIRECTION class property', () => {
      expect(Order._DEFAULT_DIRECTION).toEqual('asc');
    });

    it('instance.direction returns "asc" if no direction is specified in constructor', () => {
      const order = new Order('testProp');
      expect(order.direction).toEqual('asc');
    });

    it('instance.direction returns "desc" if direction parameter exactly matches "desc" in constructor', () => {
      const order = new Order('testProp', 'desc');
      expect(order.direction).toEqual('desc');
    });
  });

  describe('Sort class', () => {
    it('instance.toJSON() method result matches snapshot', () => {
      const sort = new Sort(orders);
      expect(sort.toJSON()).toMatchSnapshot();
    });

    it('iterates over orders using for...of loop', () => {
      const sort = new Sort(orders);
      const result = [];
      for (const order of sort) {
        result.push(order);
      }
      expect(result).toMatchSnapshot();
    });

    it('instance.forEach() method result matches snapshot', () => {
      const sort = new Sort(orders);
      const valueGroups = {
        properties: [],
        directions: [],
      };
      const iteratee = (property, direction) => {
        valueGroups.properties.push(property);
        valueGroups.directions.push(direction);
      };
      sort.forEach(iteratee);
      expect(valueGroups).toMatchSnapshot();
    });
  });

  describe('Pageable class', () => {
    it('default values match snapshot when constructor parameters are undefined', () => {
      const pageable = new Pageable();
      expect(pageable).toMatchSnapshot();
    });

    [
      { value: 'singleValue', type: 'string' },
      { value: ['valueA', 'valueB:desc', 'valueC:desc'], type: 'array of strings' },
      { value: new Sort(orders.slice(1, 3)), type: 'sort' },
    ].forEach(({ value, type }) => {
      it(`returns a valid Sort instance when a value of type ${type} is passed to the constructor`, () => {
        const pageable = new Pageable(1, 10, true, value);
        expect(pageable.sort).toMatchSnapshot();
      });
    });

    it('disregards extra commas in the sort array passed in as a parameter', () => {
      const invalidSort = ['valueA,', 'valueB,,', 'valueC,,,'];
      const result = new Pageable(0, 20, false, invalidSort);
      const invalidValues = result.sort.orders.filter(order => (order.property.length === 0));
      expect(invalidValues).toHaveLength(0);
    });

    it('disregards extra colons in the sort array passed in as a parameter', () => {
      const invalidSort = ['valueA:desc', 'valueB::desc', 'valueC::::desc'];
      const pageable = new Pageable(0, 20, false, invalidSort);
      const invalidValues = pageable.sort.orders.filter(order => (order.property.length === 0));
      expect(invalidValues).toHaveLength(0);
    });
  });

  describe('ArrayPage class', () => {
    const getValidPage = () => {
      const sort = new Sort(pageOrders);
      const pageable = new Pageable(0, 20, false, sort);
      return new ArrayPage(content, 2, pageable);
    };

    it('matches snapshot when valid parameters are passed to the constructor', () => {
      const result = getValidPage();
      expect(result).toMatchSnapshot();
    });

    it('results of instance.map() method result matches snapshot', () => {
      const page = getValidPage();
      const result = page.map((pageContent, idx) => ({ ...pageContent, age: (idx * 5) }));
      expect(result).toMatchSnapshot();
    });
  });

  describe('IndexedPage class', () => {
    const getValidIndexedPage = () => {
      const sort = new Sort(pageOrders);
      const pageable = new Pageable(0, 20, false, sort);
      const ids = [1, 2];
      const index = content.reduce((acc, record) => ({ ...acc, [record.id]: record }), {});
      return new IndexedPage(ids, index, 2, pageable);
    };

    it('matches snapshot when valid parameters are passed to the constructor', () => {
      const result = getValidIndexedPage();
      expect(result).toMatchSnapshot();
    });

    it('results of instance.map() method result matches snapshot', () => {
      const page = getValidIndexedPage();
      const result = page.map((pageContent, idx) => ({ ...pageContent, age: (idx * 5) }));
      expect(result).toMatchSnapshot();
    });
  });

  describe('IndexablePage class', () => {
    const getValidIndexablePage = (isIndexed = false) => {
      const sort = new Sort(pageOrders);
      const pageable = new Pageable(0, 20, isIndexed, sort);
      return new IndexablePage(content, 2, pageable, 'id');
    };

    it('matches snapshot when valid parameters are passed to the constructor', () => {
      const result = getValidIndexablePage();
      expect(result).toMatchSnapshot();
    });

    it('results of instance.map() method result matches snapshot', () => {
      const page = getValidIndexablePage();
      const result = page.map((pageContent, idx) => ({ ...pageContent, age: (idx * 5) }));
      expect(result).toMatchSnapshot();
    });

    it('instance.toJSON() method result matches snapshot', () => {
      const page = getValidIndexablePage(true);
      const result = page.toJSON();
      expect(result).toMatchSnapshot();
    });
  });

  describe('paginate function', () => {
    const next = () => {};
    const context = {
      query: {
        page: 1,
        size: 20,
        sort: 'firstName',
        indexed: 'true',
      },
      state: {},
    };

    it('context.state matches snapshot when paginate is called with valid context', async () => {
      await paginate(context, next);
      expect(context.state).toMatchSnapshot();
    });

    it('context.state matches snapshot when paginate is called with string values for page and size', async () => {
      const updatedQuery = { ...context.query, page: '10', size: '15' };
      await paginate({ ...context, query: updatedQuery }, next);
      expect(context.state).toMatchSnapshot();
    });

    it('context.state matches snapshot when paginate is called with empty strings for page and size', async () => {
      const updatedQuery = { ...context.query, page: '', size: '' };
      await paginate({ ...context, query: updatedQuery }, next);
      expect(context.state).toMatchSnapshot();
    });

    it('context.state matches snapshot when paginate is called with undefined page, size, and sort', async () => {
      const updatedQuery = { indexed: 'true' };
      await paginate({ ...context, query: updatedQuery }, next);
      expect(context.state).toMatchSnapshot();
    });

    it('throws error when paginate is called with invalid values for page and size', async () => {
      const updatedQuery = { ...context.query, page: 'page', size: 'size' };
      try {
        await paginate({ ...context, query: updatedQuery }, next);
      } catch (e) {
        expect(e).toBeInstanceOf(NumberFormatError);
      }
    });

    it('throws error when invalid sort direction is provided', async () => {
      const updatedQuery = { ...context.query, sort: 'firstName:foo' };
      try {
        await paginate({ ...context, query: updatedQuery }, next);
      } catch (e) {
        expect(e).toBeInstanceOf(InvalidSortError);
      }
    });
  });
});
