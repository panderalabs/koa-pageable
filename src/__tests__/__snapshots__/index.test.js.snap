// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests IndexablePage class instance.toJSON() method result matches snapshot 1`] = `
IndexedPage {
  "first": true,
  "ids": Array [
    1,
    2,
    3,
    4,
  ],
  "index": Object {
    "1": Object {
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    "2": Object {
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    "3": Object {
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    "4": Object {
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  },
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests IndexablePage class matches snapshot when valid parameters are passed to the constructor 1`] = `
ArrayPage {
  "content": Array [
    Object {
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    Object {
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    Object {
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    Object {
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  ],
  "first": true,
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests IndexablePage class results of instance.map() method result matches snapshot 1`] = `
ArrayPage {
  "content": Array [
    Object {
      "age": 0,
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    Object {
      "age": 5,
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    Object {
      "age": 10,
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    Object {
      "age": 15,
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  ],
  "first": true,
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests IndexedPage class matches snapshot when valid parameters are passed to the constructor 1`] = `
IndexedPage {
  "first": true,
  "ids": Array [
    1,
    2,
  ],
  "index": Object {
    "1": Object {
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    "2": Object {
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    "3": Object {
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    "4": Object {
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  },
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests IndexedPage class results of instance.map() method result matches snapshot 1`] = `
IndexedPage {
  "first": true,
  "ids": Array [
    1,
    2,
  ],
  "index": Object {
    "1": Object {
      "age": 5,
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    "2": Object {
      "age": 10,
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    "3": Object {
      "age": 15,
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    "4": Object {
      "age": 20,
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  },
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests ArrayPage class matches snapshot when valid parameters are passed to the constructor 1`] = `
ArrayPage {
  "content": Array [
    Object {
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    Object {
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    Object {
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    Object {
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  ],
  "first": true,
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests ArrayPage class results of instance.map() method result matches snapshot 1`] = `
ArrayPage {
  "content": Array [
    Object {
      "age": 0,
      "firstName": "Bob",
      "id": 1,
      "lastName": "Stevens",
    },
    Object {
      "age": 5,
      "firstName": "Steve",
      "id": 2,
      "lastName": "Bobbins",
    },
    Object {
      "age": 10,
      "firstName": "Robert",
      "id": 3,
      "lastName": "Stevenson",
    },
    Object {
      "age": 15,
      "firstName": "Stevarino",
      "id": 4,
      "lastName": "Robertson",
    },
  ],
  "first": true,
  "last": true,
  "number": 0,
  "numberOfElements": 4,
  "size": 20,
  "sort": Array [
    Order {
      "direction": "asc",
      "property": "firstName",
    },
    Order {
      "direction": "asc",
      "property": "lastName",
    },
  ],
  "totalElements": 2,
  "totalPages": 1,
}
`;

exports[`Tests Pageable class default values match snapshot when constructor parameters are undefined 1`] = `
Pageable {
  "indexed": false,
  "page": 0,
  "size": 20,
}
`;

exports[`Tests Pageable class returns a valid Sort instance when a value of type array of strings is passed to the constructor 1`] = `
Array [
  Order {
    "direction": "asc",
    "property": "valueA",
  },
  Order {
    "direction": "desc",
    "property": "valueB",
  },
  Order {
    "direction": "desc",
    "property": "valueC",
  },
]
`;

exports[`Tests Pageable class returns a valid Sort instance when a value of type sort is passed to the constructor 1`] = `
Array [
  Order {
    "direction": "desc",
    "property": "propertyB",
  },
  Order {
    "direction": "asc",
    "property": "propertyC",
  },
]
`;

exports[`Tests Pageable class returns a valid Sort instance when a value of type string is passed to the constructor 1`] = `
Array [
  Order {
    "direction": "asc",
    "property": "singleValue",
  },
]
`;

exports[`Tests Sort class instance.forEach() method result matches snapshot 1`] = `
Object {
  "directions": Array [
    "asc",
    "desc",
    "asc",
    "desc",
    "desc",
  ],
  "properties": Array [
    "propertyA",
    "propertyB",
    "propertyC",
    "propertyD",
    "propertyE",
  ],
}
`;

exports[`Tests Sort class instance.toJSON() method result matches snapshot 1`] = `
Array [
  Order {
    "direction": "asc",
    "property": "propertyA",
  },
  Order {
    "direction": "desc",
    "property": "propertyB",
  },
  Order {
    "direction": "asc",
    "property": "propertyC",
  },
  Order {
    "direction": "desc",
    "property": "propertyD",
  },
  Order {
    "direction": "desc",
    "property": "propertyE",
  },
]
`;

exports[`Tests Sort class iterates over orders using for...of loop 1`] = `
Array [
  Order {
    "direction": "desc",
    "property": "propertyB",
  },
  Order {
    "direction": "asc",
    "property": "propertyC",
  },
  Order {
    "direction": "desc",
    "property": "propertyD",
  },
  Order {
    "direction": "desc",
    "property": "propertyE",
  },
  undefined,
]
`;

exports[`Tests paginate function context.state matches snapshot when paginate is called with empty strings for page and size 1`] = `
Object {
  "pageable": Pageable {
    "indexed": true,
    "page": 0,
    "size": 10,
    "sort": Array [
      Order {
        "direction": "asc",
        "property": "firstName",
      },
    ],
  },
}
`;

exports[`Tests paginate function context.state matches snapshot when paginate is called with string values for page and size 1`] = `
Object {
  "pageable": Pageable {
    "indexed": true,
    "page": 10,
    "size": 15,
    "sort": Array [
      Order {
        "direction": "asc",
        "property": "firstName",
      },
    ],
  },
}
`;

exports[`Tests paginate function context.state matches snapshot when paginate is called with undefined page, size, and sort 1`] = `
Object {
  "pageable": Pageable {
    "indexed": true,
    "page": 0,
    "size": 10,
  },
}
`;

exports[`Tests paginate function context.state matches snapshot when paginate is called with valid context 1`] = `
Object {
  "pageable": Pageable {
    "indexed": true,
    "page": 1,
    "size": 20,
    "sort": Array [
      Order {
        "direction": "asc",
        "property": "firstName",
      },
    ],
  },
}
`;
